### PR TITLE
initial win support

### PR DESCRIPTION
**Pull request by isuruf**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by isuruf** [comment]
@conda-forge-admin, rerender

**Comment by isuruf** [comment]
@mattip, I don't intend to work on this anytime soon. Feel free to push to this branch.

**Comment by h-vetinari** [comment]
Cool, thanks Isuru! I've been wanting to open a PR for this for a while, but been too swamped with other things - sorry.

Still, if you want, I'd be happy to (try to) contribute to this PR if you can invite me to collaborate on your fork of this feedstock (no other way as non-core, non-feedstock-maintainer if it's not my own PR).

**Comment by mattip** [comment]
I updated the pypy code to v7.3.5 (for python3.7). Should the build number be 0 or 1?

**Comment by mattip** [comment]
Hmm. Compilation on windows runs out of memory.  According to the [azure documentation](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml#hardware) these are [Standard_DS2_v2](https://docs.microsoft.com/en-us/azure/virtual-machines/dv2-dsv2-series#dsv2-series) machines with2 vcpus and 7GB memory? Is there a way to get a beefier machine?

<details>

```
starting compile_c
[platform:execute] jom.exe -j 2 in c:\users\vssadm~1\appdata\local\temp\usession-release-pypy3.7-v7.3.5rc3-0\testing_1
Exception in thread Thread-451:
Traceback (most recent call last):
  File "D:\bld\pypy3.7_1623301055471\work\pypy2-binary\lib-python\2.7\threading.py", line 820, in __bootstrap_inner
    self.run()
  File "D:\bld\pypy3.7_1623301055471\work\pypy2-binary\lib-python\2.7\threading.py", line 773, in run
    self.__target(*self.__args, **self.__kwargs)
  File "D:\bld\pypy3.7_1623301055471\work\pypy2-binary\lib-python\2.7\subprocess.py", line 729, in _readerthread
    buffer.append(fh.read())
MemoryError
```

**Comment by h-vetinari** [comment]
> Is there a way to get a beefier machine?

It's been a problem for a while: https://github.com/conda-forge/conda-forge.github.io/issues/902, though there's some effort around getting a special CI queue off the ground: https://github.com/conda-forge/conda-forge.github.io/issues/1272

File: recipe/bld.bat [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+move %PREFIX%\README.rst %PREFIX%\lib_pypy\
+REM License is packaged separately
+del %PREFIX%\LICENSE
+
+REM Make sure the site-packages dir match with cpython
+set PY_VERSION=%name_suffix%
+mkdir  %PREFIX%\lib\python%PY_VERSION%\site-packages
+move %PREFIX%\site-packages\README %PREFIX%\lib\python%PY_VERSION%\site-packages\
+rmdir /q /s %PREFIX%\site-packages
+mklink /D %PREFIX%\site-packages %PREFIX%\lib\python%PY_VERSION%\site-packages
```
Comment (isuruf) [inline recipe/bld.bat:58]: `mklink` doesn't work without administrator privileges and conda-build doesn't support symlinks on windows AFAIK.
What we need is a way to tell PyPy that the site-packages need to be `%PREFIX%\Lib\site-packages` on windows. Relevant code https://github.com/mozillazg/pypy/blob/release-pypy3.7-v7.3.5/lib-python/3/site.py#L347-L356

Comment (isuruf) [reply recipe/bld.bat:58]: Something like,
```
diff --git a/lib-python/3/site.py b/lib-python/3/site.py
index 84de242aee..51d7a66fa3 100644
--- a/lib-python/3/site.py
+++ b/lib-python/3/site.py
@@ -344,9 +344,6 @@ def getsitepackages(prefixes=None):
             continue
         seen.add(prefix)
 
-        if is_pypy:
-            sitepackages.append(os.path.join(prefix, "site-packages"))
-            continue
         if os.sep == '/':
             sitepackages.append(os.path.join(prefix, "lib",
                                         "python%d.%d" % sys.version_info[:2],
diff --git a/lib-python/3/sysconfig.py b/lib-python/3/sysconfig.py
index d9eb7bfe79..884c7b04d5 100644
--- a/lib-python/3/sysconfig.py
+++ b/lib-python/3/sysconfig.py
@@ -44,8 +44,8 @@ _INSTALL_SCHEMES = {
     'pypy': {
         'stdlib': '{installed_base}/lib-{implementation_lower}',
         'platstdlib': '{base}/lib-{implementation_lower}',
-        'purelib': '{base}/site-packages',
-        'platlib': '{base}/site-packages',
+        'purelib': '{base}/lib/python{py_version_short}/site-packages',
+        'platlib': '{platbase}/lib/python{py_version_short}/site-packages',
         'include': '{installed_base}/include',
         'platinclude': '{installed_base}/include',
         'scripts': '{base}/bin',
@@ -64,8 +64,8 @@ _INSTALL_SCHEMES = {
     'pypy_nt': {
         'stdlib': '{installed_base}/lib-{implementation_lower}',
         'platstdlib': '{base}/lib-{implementation_lower}',
-        'purelib': '{base}/site-packages',
-        'platlib': '{base}/site-packages',
+        'purelib': '{base}/Lib/site-packages',
+        'platlib': '{base}/Lib/site-packages',
         'include': '{installed_base}/include',
         'platinclude': '{installed_base}/include',
         'scripts': '{base}/Scripts',
```

Comment (mattip) [reply recipe/bld.bat]: thanks. Should we remove the symlink on linux as well?

Comment (isuruf) [reply recipe/bld.bat]: I don't think we can do that now. We can do it for pypy3.8 or the next version that breaks ABI compat.

Comment (mattip) [reply recipe/bld.bat]: What do you think about using `pypy` in the {base}/lib/pypy{py_version_short}/site-packages like I proposed in https://github.com/python/cpython/pull/24628 (which has not been accepted yet)?

Comment (isuruf) [reply recipe/bld.bat]: The issue is that conda-build knows only about `{base}/lib/python{py_version_short}/site-packages` (unix) and `{base}/Lib/site-packages` (win)  and `noarch: python` packages are extracted into that folder. https://github.com/conda/conda/blob/33a142c16530fcdada6c377486f1c1a385738a96/conda/common/path.py#L171

Comment (mattip) [reply recipe/bld.bat]: hmm. That should be using one of the sysconfig templates, but changing it now will probably break the world. I will go with a patch for now to revert most of site.py and sysconfig.py to cpython's version like you suggest.

**Comment by mattip** [comment]
The build seems to have succeeded. Where can I download the result to see if it is usable?

**Comment by isuruf** [comment]
You can download them from here. https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=336749&view=artifacts&pathAsName=false&type=publishedArtifacts

After that, store `pypy3.7-7.3.5-hb504520_1.tar.bz2` in `<miniforge3-install-location>/conda-bld/win-64` and then run `conda create -n pypy pypy3.7 --use-local`

**Comment by mattip** [comment]
- The pickle file test added in the last commit is failing
- The `%PREFIX%/Lib/site-packages` directory is not in the final tar.gz
- The `python.exe`, `pythonw.exe`, `pypy.exe`, `pypyw.exe` are not in the final tar.gz

**Comment by isuruf** [comment]
> The pickle file test added in the last commit is failing

Right. The main reason for adding that test was to check if that was created. Running the tests created that pickle file. We need to figure out how to create those files.

File: recipe/bld.bat [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+REM Move the generic file name to somewhere that's specific to pypy
+move %PREFIX%\README.rst %PREFIX%\lib_pypy\
+REM License is packaged separately
+del %PREFIX%\LICENSE
+
+REM Make sure the site-packages dir match with cpython
+REM See patch site-and-sysconfig-conda.patch
+set PY_VERSION=%name_suffix%
+mkdir  %PREFIX%\lib\site-packages
+move %PREFIX%\site-packages\README %PREFIX%\lib\site-packages\
```
Comment (isuruf) [inline recipe/bld.bat:61]:
```suggestion
mkdir  %PREFIX%\Lib\site-packages
move %PREFIX%\site-packages\README %PREFIX%\Lib\site-packages\
```

**Comment by isuruf** [comment]
> The python.exe, pythonw.exe, pypy.exe, pypyw.exe are not in the final tar.gz

I can see that they are. In the log I see,
```
INFO (pypy3.7,python.exe): Needed DSO C:/Windows/System32/KERNEL32.dll found in $SYSROOT
```

**Comment by mattip** [comment]
cool, I must have pulled the wrong artifact

**Comment by isuruf** [comment]
I also see that `tcl/itcl4.2.0/itcl420t.dll` is in the package. Does that mean conda's tcl is not found?

File: recipe/bld.bat (hunk 2 above)
Comment (mattip) [reply recipe/bld.bat:61]: I see there is a `%SP_DIR%`, I will use that instead

**Comment by mattip** [comment]
> We need to figure out how to create those files.

I am debugging locally via `conda debug recipe` to work out a strategy.

> I also see that tcl/itcl4.2.0/itcl420t.dll is in the package. Does that mean conda's tcl is not found?

I will check this locally too.

**Comment by isuruf** [comment]
Here's the code in cpython. https://github.com/python/cpython/blob/e117c0283705943189e6b1aef668a1f68f3f00a4/Makefile.pre.in#L1631-L1634

**Comment by mattip** [comment]
Packaging (`package.py`) will try to find `tcl86t.dll` on the path, and then will copy that entire directory into the pypy archive on windows. There is no option so far to say "do not embed tcl", only an option to say "no tcl" which will then not build the c-extension interface at all. This is similar to cpython which supplies a complete tcl library.

What should I do for conda here?

**Comment by isuruf** [comment]
> What should I do for conda here?

Delete afterwards? Is PyPy looking for them in the exact place they PyPy copied it to?

**Comment by mattip** [comment]
Using the [archive from the latest build](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=337256&view=artifacts&pathAsName=false&type=publishedArtifacts), I still do not get python.exe, pythonw.exe, pypy.exe, pypyw.exe into the package. I see where [they are found](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=337256&view=logs&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=1183ba29-a0b5-5324-8463-2a49ace9e213&l=9064), but somehow they are being ignored. Is there some other metadata somewhere?

**Comment by isuruf** [comment]
Here's a direct link, https://artprodeus21.artifacts.visualstudio.com/A910fa339-c7c2-46e8-a579-7ea247548706/84710dde-1620-425b-80d0-4cf5baca359d/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2NvbmRhLWZvcmdlL3Byb2plY3RJZC84NDcxMGRkZS0xNjIwLTQyNWItODBkMC00Y2Y1YmFjYTM1OWQvYnVpbGRJZC8zMzcyNTYvYXJ0aWZhY3ROYW1lL2NvbmRhX2FydGlmYWN0c18zMzcyNTZfd2luXzY0X25hbWVfc3VmZml4My430/content?format=file&subPath=%2Fwin-64%2Fpypy3.7-7.3.5-hb504520_1.tar.bz2
and it has python.exe in it.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -11,31 +11,42 @@ package:
...
+      - patches/pre-7.3.6-0004.patch
+
+  - url: https://downloads.python.org/pypy/pypy2.7-v7.3.5-win64.zip           # [win]
+    sha256: 0b90eded11ba89a526c4288f17fff7e75000914ac071bd6d67912748ae89d761  # [win]
+    folder: pypy2-binary                                                      # [win]
 
 build:
-  number: 4
-  skip: True  # [win or name_suffix=="3.6"]
+  number: 1
```
Comment (isuruf) [inline recipe/meta.yaml:40]:
```suggestion
  number: 0
```

**Comment by mattip** [comment]
cool. I wonder what I am looking at then

**Comment by mattip** [comment]
I needed to delete everything under conda-bld/win-64 and rebuild the index. Now it is working:
- python, pythonw, pypy, pypyw, pypy3, pypy3w all work
- `pypy3 -m tkinker` opens a window, there is no `tcl` in the top-level pypy directory
- `lib\site-packages` appears in `sys.path`

but
- `conda install requests` fails with lots of conflicts (to be expected, right?)
- `pypy3 -m pip install pip` fails with `"No module named pip.__main__; 'pip' is a package and cannot be directly executed`

**Comment by mattip** [comment]
So ... what is next?

**Comment by isuruf** [comment]
@mattip, is there a pypy3.8 coming out soon? If so, we can migrate for pypy3.8 along with windows.

**Comment by mattip** [comment]
It will be a few months until we have our first 3.8.

**Comment by h-vetinari** [comment]
Amazing work @mattip, sorry for not being able to collaborate more.

**Comment by mattip** [comment]
The real credit goes to @isuruf and the conda-forge team for continuing to support this admittedly niche implementation.

**Comment by h-vetinari** [comment]
> The real credit goes to @isuruf and the conda-forge team for continuing to support this admittedly niche implementation.

That goes without saying (though I do try to say it as often as I can. ;-)).

Also: hopefully not niche much longer, now that it's becoming much more easily usable with conda! 🙃

**Pull request merged**